pkg/algorithm/sku_scheduler.go

```diff
@@ -0,0 +1,512 @@
+		levelLeafCellNum:  levelLeafCellNum,
+		cellLevels:        cellLevels,
+		crossPriorityPack: crossPriorityPack,
+	}
+}
+
+func (s *skuScheduler) Schedule(
+	podRootGroup *apiv2.PodGroupSpec,
+	priority CellPriority) (
+	placement PodGroupPlacement,
+	failedReason string) {
+
+	// sort pods in descending order by couting leaf cell number
+	s.sortPodGroup(podRootGroup)
+
+	// disable preemption first to reduce preemption, try to schedule
+	placement, failedReason = s.findCellsForPodGroup(podRootGroup, opportunisticPriority, nil, &placement)
+
+	// enable preemption if scheduling failed
+	if failedReason != "" && priority > opportunisticPriority {
+		placement, failedReason = s.findCellsForPodGroup(podRootGroup, priority, nil, &placement)
+	}
+
+	// convert cells to leaf cells in placement
+	if failedReason == "" {
+		for iter := placement.Iterator(); iter.HasNext(); {
+			cells, leafCells := iter.Next(), CellList{}
+			for _, c := range *cells {
+				currLevelCells := CellList{c}
+				for currLevelCells[0].GetLevel() > CellLevel(1) {
+					childLevelCells := CellList{}
+					for _, cc := range currLevelCells {
+						childLevelCells = append(childLevelCells, cc.GetChildren()...)
+					}
+					currLevelCells = childLevelCells
+				}
+				leafCells = append(leafCells, currLevelCells...)
+			}
+			*cells = leafCells
+		}
+	}
+
+	return placement, failedReason
+}
+
+func (s *skuScheduler) sortPodGroup(podGroup *apiv2.PodGroupSpec) {
+	sort.SliceStable(podGroup.Pods, func(i, j int) bool {
+		return s.countLeafCellNums(podGroup.Pods[i]) > s.countLeafCellNums(podGroup.Pods[j])
+	})
+	sortedPods := []apiv2.PodGroupMemberSpec{}
+	for _, p := range podGroup.Pods {
+		for i := int32(0); i < p.PodMinNumber; i++ {
+			sortedPods = append(sortedPods, p)
+		}
+	}
+	podGroup.Pods = sortedPods
+
+	sort.SliceStable(podGroup.ChildGroups, func(i, j int) bool {
+		return s.countLeafCellNums(podGroup.ChildGroups[i]) > s.countLeafCellNums(podGroup.ChildGroups[j])
+	})
+	for _, g := range podGroup.ChildGroups {
+		s.sortPodGroup(g)
+	}
+}
+
+func (s *skuScheduler) countLeafCellNums(x interface{}) int32 {
+	count := int32(0)
+	switch p := x.(type) {
+	case apiv2.PodGroupMemberSpec:
+		count = s.levelLeafCellNum[s.cellLevels[p.CellsPerPod.CellType]] * p.CellsPerPod.CellNumber
+	case []apiv2.PodGroupMemberSpec:
+		for _, pp := range p {
+			count += s.countLeafCellNums(pp)
+		}
+	case *apiv2.PodGroupSpec:
+		count += s.countLeafCellNums(p.Pods) + s.countLeafCellNums(p.ChildGroups)
+	case []*apiv2.PodGroupSpec:
+		for _, pp := range p {
+			count += s.countLeafCellNums(pp)
+		}
+	}
+	return count
+}
+
+func (s *skuScheduler) findCellsForPodGroup(
+	podGroup *apiv2.PodGroupSpec,
+	priority CellPriority,
+	within *skuCell,
+	allocated *PodGroupPlacement) (
+	placement PodGroupPlacement,
+	failedReason string) {
+
+	placement, failedReason = PodGroupPlacement{}, "no matched cells in vc"
+
+	cv := skuClusterView{nil}
+	if level, ok := s.cellLevels[podGroup.WithinOneCell]; ok {
```

Comment:
What will happen if withinOneCell is not set or set to "" (like [hived_algorithm_v2_test.go)](pkg/algorithm/hived_algorithm_v2_test.go)?

Reply:
cluster view will be a slice with one nil element, the following scheduling will ignore within cell if within is nil